core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1158,9 +1158,14 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
     @throws(classOf[Exception])
     def handleNewSession() {
       info("ZK expired; shut down all controller components and try to re-elect")
-      onControllerResignation()
-      inLock(controllerContext.controllerLock) {
-        controllerElector.elect
+      if (controllerElector.getControllerID() != config.brokerId) {
+        onControllerResignation()
+        inLock(controllerContext.controllerLock) {
+          controllerElector.elect
+        }
+      } else {
```

Comment:
@pengwei-li : Sorry for the delay in response. Good point that you raised and I think there is indeed an issue. If a broker is the controller and its ZK session expires, the subsequent events should be (1) new session is created; (2) KafkaController.handleNewSession() is called; (3) either LeaderChangeListener.handleDataChange or handleDataDeleted is called (depending on whether the new session has expired or not). As you can see, since handleNewSession() is called before handleDataDeleted(), I think the controller path can only be created by handleNewSession(). However, when there are successive session expirations, it's possible that the controller path is created on the newest session in the first handleNewSession(). When handling the second handleNewSession(), the controller will hit the following code and return without calling the onBecomingLeader() logic. Now the controller is in the resigned state and can't recover from it. 

```
if(leaderId != -1) {
     debug("Broker %d has been elected as leader, so stopping the election process.".format(leaderId))
     return amILeader
}
```  
Instead of fixing here, perhaps a simpler fix is to remove the following code in ZookeeperLeaderElector.elect()?
```
   leaderId = getControllerID 
    /* 
     * We can get here during the initial startup and the handleDeleted ZK callback. Because of the potential race condition, 
     * it's possible that the controller has already been elected when we get here. This check will prevent the following 
     * createEphemeralPath method from getting into an infinite loop if this broker is already the controller.
     */
    if(leaderId != -1) {
       debug("Broker %d has been elected as leader, so stopping the election process.".format(leaderId))
       return amILeader
    }
```  
I thought KAFKA-1387 fixed the multiple session expiration issue, but it seems that it didn't.